src/Toml/Bi/Combinators.hs

```diff
@@ -0,0 +1,146 @@
+
+-- | Specialied for 'Toml' monad.
+type BiToml a = Bi Env St a
+
+-- | Convert textual representation of toml into user data type.
+encode :: BiToml a -> Text -> Either EncodeException a
+encode biToml text = do
+    toml <- first ParseError (parse text)
+    runReader (runExceptT $ biRead biToml) toml
+
+-- | Convert object to textual representation.
+decode :: BiToml a -> a -> Either DecodeException Text
+decode biToml obj = do
+    -- this pair has type (TOML, Either DecodeException a)
+    let (result, toml) = runState (runExceptT $ biWrite biToml obj) (TOML mempty mempty)
+
+    -- just to trigger error if Left
+    _ <- result
+
+    pure $ prettyToml toml
+
+-- | Unsafe version of 'decode' function if you're sure that you decoding
+-- structure is correct.
+unsafeDecode :: BiToml a -> a -> Text
+unsafeDecode biToml text = fromRight (error "Unsafe decode") $ decode biToml text
+
+-- | General function to create bidirectional converters for values.
+bijectionMaker :: forall a t .
+                 Text                              -- ^ Name of expected type
+               -> (forall f . Value f -> Maybe a)  -- ^ How to convert from 'AnyValue' to @a@
+               -> (a -> Value t)                   -- ^ Convert @a@ to 'Anyvale'
+               -> Key                              -- ^ Key of the value
+               -> BiToml a
+bijectionMaker typeTag fromVal toVal key = Bijection input output
+  where
+    input :: Env a
+    input = do
+        mVal <- asks $ HashMap.lookup key . tomlPairs
+        case mVal of
+            Nothing -> throwError $ KeyNotFound key
+            Just (AnyValue val) -> case fromVal val of
+                Just v  -> pure v
+                Nothing -> throwError $ TypeMismatch typeTag
+
+    output :: a -> St a
+    output a = do
+        let val = AnyValue (toVal a)
+        mVal <- gets $ HashMap.lookup key . tomlPairs
+        case mVal of
+            Nothing -> a <$ modify (\(TOML vals nested) -> TOML (HashMap.insert key val vals) nested)
+            Just _  -> throwError $ DuplicateKey key val
+
+-- | Parser for boolean values.
+bool :: Key -> BiToml Bool
+bool = bijectionMaker "Boolean" fromBool Bool
+  where
+    fromBool :: Value f -> Maybe Bool
+    fromBool (Bool b) = Just b
+    fromBool _        = Nothing
+
+-- | Parser for integer values.
+int :: Key -> BiToml Int
+int = bijectionMaker "Int" fromInt (Int . toInteger)
+  where
+    fromInt :: Value f -> Maybe Int
+    fromInt (Int n) = Just (fromIntegral n)
+    fromInt _       = Nothing
+
+-- | Parser for floating values.
+double :: Key -> BiToml Double
+double = bijectionMaker "Double" fromDouble Float
+  where
+    fromDouble :: Value f -> Maybe Double
+    fromDouble (Float f) = Just f
+    fromDouble _         = Nothing
+
+-- | Parser for string values.
+str :: Key -> BiToml Text
+str = bijectionMaker "String" fromString String
+  where
+    fromString :: Value f -> Maybe Text
+    fromString (String s) = Just s
```

Comment:
Unfortunately, patterns are not first class objects in Haskell at the moment, so I need to write this function manually each time...